File: Harvest/Models/SafetyModels.swift
import Foundation
import SwiftUI

struct SafetyAnalysis: Codable, Identifiable, Sendable {
    let id: String
    let conversationId: String
    let userId: String
    let matchId: String
    var safetyScore: Int
    var redFlags: [SafetyFlagSnapshot]
    var recommendations: [String]
    var allowContactSharing: Bool
    var createdAt: String?

    // Derived app-side fields populated by the service.
    var otherUserId: String = ""
    var totalMessages: Int = 0
    var firstMessageAt: String?

    enum CodingKeys: String, CodingKey {
        case id
        case conversationId = "conversation_id"
        case userId = "user_id"
        case matchId = "match_id"
        case safetyScore = "safety_score"
        case redFlags = "red_flags"
        case recommendations
        case allowContactSharing = "allow_contact_sharing"
        case createdAt = "created_at"
    }

    var redFlagCount: Int {
        redFlags.count
    }

    var has24HourHistory: Bool {
        let reference = firstMessageAt ?? createdAt
        guard let reference,
              let date = ISO8601DateFormatter().date(from: reference) ??
                  ISO8601DateFormatter.fractionalSeconds.date(from: reference)
        else {
            return false
        }

        return Date().timeIntervalSince(date) >= 24 * 60 * 60
    }

    var safetyLevel: SafetyLevel {
        if safetyScore < 20 { return .block }
        if safetyScore < 50 { return .warning }
        if safetyScore < 70 { return .caution }
        if safetyScore < 80 { return .safe }
        return .verified
    }
}

struct SafetyFlagSnapshot: Codable, Identifiable, Sendable {
    let id: String
    let category: RedFlagCategory
    let severity: RedFlagSeverity
    let evidence: String
    let messageId: String?
    let createdAt: String?

    enum CodingKeys: String, CodingKey {
        case id, category, severity, evidence
        case messageId = "message_id"
        case createdAt = "created_at"
    }
}

enum SafetyLevel: String, Sendable {
    case block, warning, caution, safe, verified

    var color: Color {
        switch self {
        case .block: return Color(hex: "DC2626")
        case .warning: return Color(hex: "F59E0B")
        case .caution: return Color(hex: "F97316")
        case .safe: return Color(hex: "27CF8A")
        case .verified: return Color(hex: "3B82F6")
        }
    }

    var icon: String {
        switch self {
        case .block: return "xmark.shield.fill"
        case .warning: return "exclamationmark.shield.fill"
        case .caution: return "exclamationmark.triangle.fill"
        case .safe: return "checkmark.shield.fill"
        case .verified: return "checkmark.seal.fill"
        }
    }

    var displayName: String {
        rawValue.capitalized
    }
}

struct RedFlagReport: Codable, Identifiable, Sendable {
    let id: String
    let reporterId: String?
    let reportedUserId: String?
    let conversationId: String?
    let flagType: String
    let severity: RedFlagSeverity?
    let evidence: String?
    let aiDetected: Bool
    let userReported: Bool
    let reviewed: Bool
    let actionTaken: String?
    let reviewedBy: String?
    let reviewedAt: String?
    let createdAt: String?

    enum CodingKeys: String, CodingKey {
        case id
        case reporterId = "reporter_id"
        case reportedUserId = "reported_user_id"
        case conversationId = "conversation_id"
        case flagType = "flag_type"
        case severity
        case evidence
        case aiDetected = "ai_detected"
        case userReported = "user_reported"
        case reviewed
        case actionTaken = "action_taken"
        case reviewedBy = "reviewed_by"
        case reviewedAt = "reviewed_at"
        case createdAt = "created_at"
    }

    var category: RedFlagCategory {
        RedFlagCategory(rawValue: flagType) ?? .spam
    }

    var detail: String {
        evidence ?? ""
    }
}

enum RedFlagCategory: String, Codable, Sendable {
    case financial
    case personalInfo = "personal_info"
    case catfishing
    case manipulation
    case harassment
    case inappropriate
    case spam

    var weight: Int {
        severity.weight
    }

    var severity: RedFlagSeverity {
        switch self {
        case .financial: return .critical
        case .personalInfo: return .critical
        case .catfishing: return .high
        case .manipulation: return .medium
        case .harassment: return .high
        case .inappropriate: return .medium
        case .spam: return .low
        }
    }
}

enum RedFlagSeverity: String, Codable, Sendable {
    case low
    case medium
    case high
    case critical

    var weight: Int {
        switch self {
        case .low: return 10
        case .medium: return 20
        case .high: return 25
        case .critical: return 30
        }
    }
}

private extension ISO8601DateFormatter {
    static let fractionalSeconds: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter
    }()
}

File: android/app/src/main/java/com/harvestglass/harvest/data/model/SafetyModels.kt
package com.harvestglass.harvest.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.time.Duration
import java.time.Instant

/** Mirrors RedFlagSeverity in Harvest/Models/SafetyModels.swift. */
@Serializable
enum class RedFlagSeverity(val raw: String, val weight: Int) {
    @SerialName("low") LOW("low", 10),
    @SerialName("medium") MEDIUM("medium", 20),
    @SerialName("high") HIGH("high", 25),
    @SerialName("critical") CRITICAL("critical", 30);

    companion object {
        fun fromRaw(raw: String?): RedFlagSeverity =
            entries.firstOrNull { it.raw == raw } ?: LOW
    }
}

/** Mirrors RedFlagCategory. The severity per category is fixed, not stored. */
@Serializable
enum class RedFlagCategory(val raw: String, val severity: RedFlagSeverity) {
    @SerialName("financial") FINANCIAL("financial", RedFlagSeverity.CRITICAL),
    @SerialName("personal_info") PERSONAL_INFO("personal_info", RedFlagSeverity.CRITICAL),
    @SerialName("catfishing") CATFISHING("catfishing", RedFlagSeverity.HIGH),
    @SerialName("manipulation") MANIPULATION("manipulation", RedFlagSeverity.MEDIUM),
    @SerialName("harassment") HARASSMENT("harassment", RedFlagSeverity.HIGH),
    @SerialName("inappropriate") INAPPROPRIATE("inappropriate", RedFlagSeverity.MEDIUM),
    @SerialName("spam") SPAM("spam", RedFlagSeverity.LOW);

    val weight: Int get() = severity.weight

    companion object {
        /** Unknown types read as spam, matching Swift's `?? .spam`. */
        fun fromRaw(raw: String?): RedFlagCategory =
            entries.firstOrNull { it.raw == raw } ?: SPAM
    }
}

/** Mirrors SafetyLevel. Colours live in the UI layer, as they do on iOS. */
enum class SafetyLevel(val raw: String) {
    BLOCK("block"),
    WARNING("warning"),
    CAUTION("caution"),
    SAFE("safe"),
    VERIFIED("verified");

    val displayName: String get() = raw.replaceFirstChar { it.uppercase() }
}

@Serializable
data class SafetyFlagSnapshot(
    val id: String,
    val category: RedFlagCategory,
    val severity: RedFlagSeverity,
    val evidence: String,
    @SerialName("message_id") val messageId: String? = null,
    @SerialName("created_at") val createdAt: String? = null
)

/** Mirrors SafetyAnalysis in Harvest/Models/SafetyModels.swift. */
@Serializable
data class SafetyAnalysis(
    val id: String,
    @SerialName("conversation_id") val conversationId: String,
    @SerialName("user_id") val userId: String,
    @SerialName("match_id") val matchId: String,
    @SerialName("safety_score") val safetyScore: Int = 100,
    @SerialName("red_flags") val redFlags: List<SafetyFlagSnapshot> = emptyList(),
    val recommendations: List<String> = emptyList(),
    @SerialName("allow_contact_sharing") val allowContactSharing: Boolean = false,
    @SerialName("created_at") val createdAt: String? = null
) {
    // Derived app-side fields, populated by the service rather than the table.
    @kotlinx.serialization.Transient
    var otherUserId: String = ""

    @kotlinx.serialization.Transient
    var totalMessages: Int = 0

    @kotlinx.serialization.Transient
    var firstMessageAt: String? = null

    val redFlagCount: Int get() = redFlags.size

    /**
     * Whether the conversation is at least a day old. Falls back to the
     * analysis row's own timestamp when no message has been sent yet.
     */
    val has24HourHistory: Boolean
        get() {
            val reference = firstMessageAt ?: createdAt ?: return false
            val started = runCatching { Instant.parse(reference) }.getOrNull() ?: return false
            return Duration.between(started, Instant.now()).toHours() >= 24
        }

    val safetyLevel: SafetyLevel
        get() = when {
            safetyScore < 20 -> SafetyLevel.BLOCK
            safetyScore < 50 -> SafetyLevel.WARNING
            safetyScore < 70 -> SafetyLevel.CAUTION
            safetyScore < 80 -> SafetyLevel.SAFE
            else -> SafetyLevel.VERIFIED
        }
}

@Serializable
data class RedFlagReport(
    val id: String,
    @SerialName("reporter_id") val reporterId: String? = null,
    @SerialName("reported_user_id") val reportedUserId: String? = null,
    @SerialName("conversation_id") val conversationId: String? = null,
    @SerialName("flag_type") val flagType: String,
    val severity: RedFlagSeverity? = null,
    val evidence: String? = null,
    @SerialName("ai_detected") val aiDetected: Boolean = false,
    @SerialName("user_reported") val userReported: Boolean = false,
    val reviewed: Boolean = false,
    @SerialName("action_taken") val actionTaken: String? = null,
    @SerialName("reviewed_by") val reviewedBy: String? = null,
    @SerialName("reviewed_at") val reviewedAt: String? = null,
    @SerialName("created_at") val createdAt: String? = null
) {
    val category: RedFlagCategory get() = RedFlagCategory.fromRaw(flagType)

    val detail: String get() = evidence.orEmpty()
}
